---
title: "Untitled"
author: "Alessadnro Salvatore Raho"
date: "5/10/2023"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("ggplot2")
library("biomaRt")
library("edgeR")
library("tidyverse")
library("GenomicFeatures")
library("clusterProfiler")
library("org.Hs.eg.db")
library("enrichplot")
library("DOSE")
library("pathview")
library("ggnewscale")
library("fgsea")
library("PWMEnrich")
library("PWMEnrich.Hsapiens.background")
library("MotifDb")
library("igraph")
```

**First: load data**

```{r}
setwd("~/Desktop/")
load("~/Downloads/Lung_squamous_cell_carcinoma.RData")
```
Loaded the data we can inspect the three datasets gathering some basic information.
```{r}
dim(raw_counts_df)
head(raw_counts_df)
dim(r_anno_df)
head(r_anno_df)
dim(c_anno_df)
head(c_anno_df)
```

```{r}
ensembl<- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
anno_query<- getBM(attributes = c("ensembl_gene_id", "gene_biotype"), filters = ("ensembl_gene_id"), values = r_anno_df$ensembl_gene_id, mart = ensembl)
dim(anno_query)
head(anno_query)

coding_query<-anno_query[which(anno_query$gene_biotype == "protein_coding"), ]

tmp<-raw_counts_df
coding_raw<- merge(coding_query,tmp, by.y = 'row.names', by.x = "ensembl_gene_id")
clean_coding_raw<- coding_raw[1:nrow(coding_raw),3:ncol(coding_raw)]
row.names(clean_coding_raw)= coding_raw$ensembl_gene_id
```

```{r}
size_raw<-data.frame("sample"=colnames(tmp2),"read_millions"=colSums(clean_coding_raw/1000000))
ggplot(data=size_raw, aes(sample,read_millions))+
  geom_bar(stat="identity", fill="grey50",colour="grey50", width=0.7, alpha=0.7) + coord_flip() + theme_bw()
```

```{r}
threshold<-20
repl_thr<-5
filter_vec <- apply(tmp2,1,
    function(y) max(by(y, c_anno_df$condition, function(x) sum(x>=threshold))))
# see statistics for the filtering
table(filter_vec) 
filter_counts_df <- clean_coding_raw[filter_vec>=repl_thr,]
# check the dimension of the filtered matrix 
dim(filter_counts_df) 
```

```{r}
filter_anno_df <- r_anno_df[rownames(filter_counts_df),]
dim(filter_anno_df)
long_counts_df <- gather(filter_counts_df, key = "sample", value = "read_number")

# plot in log10 scale +1 to avoid infinite values
ggplot(data=long_counts_df,aes(sample,read_number+1)) +
   geom_boxplot(colour="indianred",fill="indianred",alpha=0.7) +
   theme_bw() +
   scale_y_log10()

```

```{r}
size_df <- data.frame("sample"=colnames(filter_counts_df),
                    "read_millions"=colSums(filter_counts_df)/1000000)

ggplot(data=size_df,aes(sample,read_millions)) +
   geom_bar(stat="identity",fill="indianred",colour="indianred",width=0.7,alpha=0.7)+
   coord_flip()+
   theme_bw()
```

```{r}
# scale for rows (each gene), to normalize for basal expression level differences
clu_data <- scale(t(filter_counts_df))
# distance matrix (we are clustering samples, not genes)
dd <- dist(clu_data, method = "euclidean") 
hc <- hclust(dd, method="ward.D")
# display dendogram
plot(hc) 

## PCA analysis of samples
data.matrix <- filter_counts_df
#data.matrix_t = t(data.matrix)
#ordered_matrix_t<- merge(c_anno_df, data.matrix_t, by.y = 'row.names', by.x = "sample")
#sorted_ordered_matrix_t <- ordered_matrix_t[order(ordered_matrix_t$condition),]

#color <- c(rep("red",50),rep("blue",50))
color <- c(rep("red",100))
color[grep('11A',colnames(data.matrix))]<-'blue'
#row.names(sorted_ordered_matrix_t)= sorted_ordered_matrix_t$sample
#matrix <- sorted_ordered_matrix_t[1:nrow( sorted_ordered_matrix_t),3:ncol(sorted_ordered_matrix_t)]


data.PC <- prcomp(t(data.matrix),scale.=TRUE)
plot(data.PC$x[,1:2],xlim=c(-200,200),ylim=c(-200,200),col=color,pch=19)
text(data.PC$x[,1],data.PC$x[,2]+10,colnames(filter_counts_df),cex=0.7)
```

```{r}
#c_anno_df_copy <- c_anno_df
#c_anno_df <- c_anno_df[order(ordered_matrix_t$condition),]
#f_matrix <- t(matrix)
edge_c <- DGEList(counts=filter_counts_df,group=c_anno_df$condition,samples=c_anno_df,genes=filter_anno_df) 
edge_c
```

```{r}
edge_n <- calcNormFactors(edge_c,method="TMM")
edge_n
```

```{r}
cpm_table <- as.data.frame(round(cpm(edge_n),2))
head(cpm_table)
long_cpm_df <- gather(cpm_table, key = "sample", value = "CPM") 

ggplot(data=long_cpm_df,aes(sample,CPM+1)) +
   geom_boxplot(colour="olivedrab",fill="olivedrab",alpha=0.7)+
   theme_bw()+
   scale_y_log10() 
```

```{r}
clu_data <- t(scale(t(cpm_table))) 
dd <- dist(t(clu_data), method = "euclidean") 
hc <- hclust(dd, method="ward.D")
plot(hc) 
```

```{r}
data.matrix <- cpm_table
color <- c(rep("red",100))
color[grep('11A',colnames(data.matrix))]<-'blue'
data.PC <- prcomp(t(data.matrix),scale.=TRUE)
plot(data.PC$x[,1:2],xlim=c(-200,200),ylim=c(-200,200),col=color,pch=19)
text(data.PC$x[,1],data.PC$x[,2]+10,colnames(filter_counts_df),cex=0.7)
```

```{r}
design <- model.matrix(~0+group, data=edge_n$samples)
colnames(design) <- levels(edge_n$samples$group)
rownames(design) <- edge_n$samples$sample
design
```

```{r}
# calculate dispersion and fit with edgeR (necessary for differential expression analysis)
edge_d <- estimateDisp(edge_n,design)
edge_f <- glmQLFit(edge_d,design) 
```

```{r}
contro <- makeContrasts("case-control", levels=design) 
```

```{r}
# fit the model with generalized linear models
edge_t <- glmQLFTest(edge_f,contrast=contro)
DEGs <- as.data.frame(topTags(edge_t,n=20,p.value = 0.01,sort.by = "logFC"))
DEGs <- as.data.frame(topTags(edge_t,n=20000))
DEGs$class <- "="
DEGs$class[which(DEGs$logCPM>1&DEGs$logFC>1.5&DEGs$FDR<0.01)] = "+"
DEGs$class[which(DEGs$logCPM>1&DEGs$logFC<(-1.5)&DEGs$FDR<0.01)] = "-"
DEGs <- DEGs[order(DEGs$logFC,decreasing = T),]

head(DEGs)
table(DEGs$class)
```

```{r}
## Display an MA plot
input_df <- DEGs
xlabel <- "log2 avg CPM (A)"
ylabel <- "log2 FC case vs control (M)"
par(fig=c(0,1,0,1), mar=c(4,4,1,2), mgp=c(2, 0.75, 0))	
plot(input_df$logCPM, input_df$logFC, xlab=xlabel, ylab=ylabel, 
     col=ifelse(input_df$class=="=","grey70","olivedrab4"), pch=20, frame.plot=TRUE, cex=0.8, main="MA plot")
abline(h=0,lty=2,col="grey20")
```

```{r}
## Display a Volcano plot of the results:
input_df <- DEGs
xlabel <- "log2 FC case vs control"
ylabel <- "-log10 p-value"
par(fig=c(0,1,0,1), mar=c(4,4,1,2), mgp=c(2, 0.75, 0))	
plot(input_df$logFC, -log(input_df$PValue,base=10),xlab=xlabel, ylab=ylabel, 
     col=ifelse(input_df$class=="=","grey70","olivedrab4"), pch=20, frame.plot=TRUE, cex=0.8, main="Volcano plot")
abline(v=0,lty=2,col="grey20")

```

```{r}
#data.matrix <- cpm_table
#data.matrix_t = t(data.matrix)

#ordered_matrix_t<- merge(c_anno_df, data.matrix_t, by.y = 'row.names', by.x = "sample")
#sorted_ordered_matrix_t <- ordered_matrix_t[order(ordered_matrix_t$condition),]
#row.names(sorted_ordered_matrix_t)= sorted_ordered_matrix_t$sample
#matrix <- sorted_ordered_matrix_t[1:nrow( sorted_ordered_matrix_t),3:ncol(sorted_ordered_matrix_t)]
#t_cpm_table <- t(matrix)


cols <- c(rep("chartreuse4",100)) 
cols[grep('11A',colnames(cpm_table))]<-'burlywood3'
pal <- c("blue","white","red") 
pal <- colorRampPalette(pal)(100)
sub_matrix <- as.matrix(cpm_table[which(rownames(cpm_table)%in%DEGs$ensembl_gene_id[which(DEGs$class!="=")]),])
heatmap(sub_matrix, ColSideColors = cols,cexCol = 0.5,margins = c(4,4),col=pal,cexRow = 0.2)

#heatmap(as.matrix(cpm_table[which(rownames(cpm_table)%in%DEGs$ensembl_gene_id[which(DEGs$class!="=")]),]), ColSideColors = cols,cexCol = 0.5,margins = c(4,4),col=pal,cexRow = 0.2)
```

```{r}
### Use biomaRt to map Gene symbols, Entrez IDs and Ensembl gen IDs
convert <- getBM(attributes=c("ensembl_gene_id","entrezgene_id","external_gene_name"),
                 filters=c("ensembl_gene_id"), 
                 values=DEGs$ensembl_gene_id,
                 mart = ensembl)
DEGs <- merge(DEGs,convert,by.x="ensembl_gene_id",by.y="ensembl_gene_id")
DEGs <- DEGs[1:nrow(DEGs),1:(ncol(DEGs)-1)]
colnames(DEGs)[2] <- "external_gene_name"
DEGs <- DEGs[which(!is.na(DEGs$entrezgene_id)),] #17741 vs 17355
DEGs <- DEGs[-which(duplicated(DEGs$entrezgene_id)),] # 17275
```


```{r}
ranks <- DEGs$logFC
names(ranks) <- DEGs$entrezgene_id
head(ranks)
barplot(sort(ranks, decreasing = T), las=3)
```

```{r}
## Create a list of upregulated genes
upDEGs <- DEGs %>% filter(class == "+")
```

```{r}
## Perform Gene Ontology enrichment analysis (Biological Process)
ego_BP_up <- enrichGO(gene = upDEGs$external_gene_name,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "BP",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```

```{r}
## Visualize the top 10 enriched terms with a barplot 
barplot(ego_BP_up, showCategory = 10)
dotplot(ego_BP_up, showCategory=10)
heatplot(ego_BP_up,showCategory =10)
```

```{r}
x_BP_up<- pairwise_termsim(ego_BP_up)
emapplot(x_BP_up,showCategory = 10)
```

```{r}
## Create a list of upregulated genes
downDEGs <- DEGs %>% filter(class == "-")
```

```{r}
## Perform Gene Ontology enrichment analysis (Biological Process)
ego_BP_down <- enrichGO(gene = downDEGs$external_gene_name,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "BP",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```

```{r}
## Visualize the top 10 enriched terms with a barplot 
barplot(ego_BP_down, showCategory = 10)
dotplot(ego_BP_down, showCategory=10)
heatplot(ego_BP_down,showCategory =10)
```

```{r}
x_BP_down<- pairwise_termsim(ego_BP_down)
emapplot(x_BP_down,showCategory = 10)
```
```{r}
## Perform Gene Ontology enrichment analysis (Biological Process)
ego_MF_up <- enrichGO(gene = upDEGs$external_gene_name,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "MF",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```

```{r}
## Visualize the top 10 enriched terms with a barplot 
barplot(ego_MF_up, showCategory = 10)
dotplot(ego_MF_up, showCategory=10)
heatplot(ego_MF_up,showCategory =10)
```

```{r}
x_MF_up<- pairwise_termsim(ego_MF_up)
emapplot(x_MF_up,showCategory = 10)
```

```{r}
## Perform Gene Ontology enrichment analysis (Biological Process)
ego_MF_down <- enrichGO(gene = downDEGs$external_gene_name,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "MF",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```

```{r}
## Visualize the top 10 enriched terms with a barplot 
barplot(ego_MF_down, showCategory = 10)
dotplot(ego_MF_down, showCategory=10)
heatplot(ego_MF_down,showCategory =10)
```

```{r}
x_MF_down<- pairwise_termsim(ego_MF_down)
emapplot(x_MF_down,showCategory = 10)
```

```{r}
eWP_up = enrichWP(gene = upDEGs$entrezgene_id,
                  organism = "Homo sapiens",
                  pvalueCutoff = 0.05,
                  qvalueCutoff = 0.05)
head(eWP_up, n=10)
```

```{r}
eWP_down = enrichWP(gene = downDEGs$entrezgene_id,
                  organism = "Homo sapiens",
                  pvalueCutoff = 0.05,
                  qvalueCutoff = 0.05)
head(eWP_down, n=10)
```

```{r}
logFC <- upDEGs$logFC
names(logFC) <- upDEGs$entrezgene_id
pathview(gene.data = logFC,
         pathway.id = 'hsa05222',
         species = "human")
```

```{r}
kegg <- enrichKEGG(
  gene = upDEGs$entrezgene_id,
  organism = "hsa",
  keyType = "kegg",
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  minGSSize = 10,
  maxGSSize = 500,
  qvalueCutoff = 0.05,
  use_internal_data = FALSE
)
head(kegg, n=10)
```
```{r}
logFC_ex <- upDEGs$logFC
names(logFC_ex) <- upDEGs$entrezgene_id
pathview(gene.data = logFC,
         pathway.id = 'hsa04110',
         species = "human")
```



```{r}
ensembl <-useMart(biomart = 'ensembl', dataset = 'hsapiens_gene_ensembl')
promoter_seq <- getSequence(id = upDEGs$ensembl_gene_id,
                            type = 'ensembl_gene_id',
                            seqType = 'gene_flank',
                            upstream = 500,
                            mart = ensembl
)

data("PWMLogn.hg19.MotifDb.Hsap")
sequences <- lapply(promoter_seq$gene_flank, function(x) DNAString(x))
enriched_TFs <- motifEnrichment(sequences, PWMLogn.hg19.MotifDb.Hsap, score = "affinity")
report <- groupReport(enriched_TFs, by.top.motifs = T)
report[report$p.value < 0.05]
plot(report[1:5], fontsize = 7, id.fontsize=5)

```


```{r}
TF <- report$target[2]
TF_motifs <- subset(MotifDb, organism == 'Hsapiens' & geneSymbol == TF)
PWM = toPWM(as.list(TF_motifs))
names(PWM) = sapply(names(PWM),function(x) strsplit(x,"-")[[1]][3])

# raw.scores = T list of raw score values before cutoff
scores<- motifScores(sequences,PWM, raw.scores = T, verbose=T)
ecdf <- motifEcdf(PWM,organism = "hg19",quick=TRUE)
threshold <- lapply(ecdf, function(x) quantile(x, 0.999))
scores = motifScores(sequences, PWM, raw.scores = FALSE, cutoff = unlist(threshold))
scores_sign <- which(apply(scores,1,sum) > 0)
enriched_jund <- upDEGs[scores_sign, 8]
#############################

```

```{r}
upreg_string <- upDEGs[order(upDEGs$PValue),]
downreg_string <- downDEGs[order(downDEGs$PValue),]
diff_string <- rbind(upreg_string,downreg_string)
# All differentially expressed genes
write(diff_string$ensembl_gene_id,"degs_IDs.txt")
write(upreg_string$ensembl_gene_id[1:100], "upreg_IDs.txt")
write(downreg_string$ensembl_gene_id[1:100], "downreg_IDs.txt")
write(rbind(upreg_string$ensembl_gene_id[1:100],downreg_string$ensembl_gene_id[1:100]), "degs_IDs_200.txt")
```

```{r}
links_degs <-  read.delim("string_interactions_all_200.tsv")
links_up <- read.delim("string_interactions_up100.tsv")
links_down <- read.delim("string_interactions_down_100.tsv")

draw_largest_comp <- function(links){
  nodes<-union(links[,1],links[,2])
  net <- graph_from_data_frame(d=links,vertices=nodes,directed=FALSE)
  comp <- components(net, mode = "strong")
  # find largest component
  biggest_comp <- which.max(comp$csize)
  # print largest component size ## 1676
  print(comp$csize[biggest_comp])
  # isolate largest component
  first_c<-induced_subgraph(net,V(net)[comp$membership == biggest_comp])
  plot(first_c,
       # edge proportional to combined score
       edge.width=E(first_c)$combined_score*3,
       vertex.color="orange",
       vertex.size=10,
       vertex.frame.color="darkgray",
       vertex.label.color="black",
       vertex.label.cex=0.7,
       edge.curved=0.1)
}
draw_largest_comp(links_degs)
draw_largest_comp(links_up)
draw_largest_comp(links_down)
```