---
title: "Bioinformatic Resources Project"
author: "Camilla Pelagalli, Alessadnro Salvatore Raho"
date: "5/2/2023"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("ggplot2")
library("biomaRt")
library("edgeR")
library("tidyverse")
library("GenomicFeatures")
library("clusterProfiler")
library("org.Hs.eg.db")
library("enrichplot")
library("DOSE")
library("pathview")
library("ggnewscale")
library("fgsea")
```

punto 1: load data

```{r}
setwd("~/Desktop/")
load("~/Desktop/Bioinformatic-resources-Project-main/Lung_squamous_cell_carcinoma.RData")
```

Check number of transcripts

```{r}
dim(raw_counts_df)
head(raw_counts_df)
dim(r_anno_df)
head(r_anno_df)
dim(c_anno_df)
head(c_anno_df)
```

punto 2: update data extracting only coding regions

```{r}
ensembl<- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
anno_query<- getBM(attributes = c("ensembl_gene_id", "gene_biotype"), filters = ("ensembl_gene_id"), values = r_anno_df$ensembl_gene_id, mart = ensembl)
dim(anno_query)
head(anno_query)

coding_query<-anno_query[which(anno_query$gene_biotype == "protein_coding"), ]

tmp<-raw_counts_df
coding_raw<- merge(coding_query,tmp, by.y = 'row.names', by.x = "ensembl_gene_id")
tmp2<- coding_raw[1:nrow(coding_raw),3:ncol(coding_raw)]
row.names(tmp2)= coding_raw$ensembl_gene_id
```

```{r}

```

Check the library size of each sample

```{r}
size_raw<-data.frame("sample"=colnames(tmp2),"read_millions"=colSums(tmp2/1000000))
ggplot(data=size_raw, aes(sample,read_millions))+
  geom_bar(stat="identity", fill="grey50",colour="grey50", width=0.7, alpha=0.7) + coord_flip() + theme_bw()
```

```{r}
threshold<-20
repl_thr<-5
filter_vec <- apply(tmp2,1,
    function(y) max(by(y, c_anno_df$condition, function(x) sum(x>=threshold))))
# see statistics for the filtering
table(filter_vec) 
filter_counts_df <- tmp2[filter_vec>=repl_thr,]
# check the dimension of the filtered matrix 
dim(filter_counts_df) 
```

```{r}
filter_anno_df <- r_anno_df[rownames(filter_counts_df),]
dim(filter_anno_df)
long_counts_df <- gather(filter_counts_df, key = "sample", value = "read_number")

# plot in log10 scale +1 to avoid infinite values
ggplot(data=long_counts_df,aes(sample,read_number+1)) +
   geom_boxplot(colour="indianred",fill="indianred",alpha=0.7) +
   theme_bw() +
   scale_y_log10()

```

```{r}
size_df <- data.frame("sample"=colnames(filter_counts_df),
                    "read_millions"=colSums(filter_counts_df)/1000000)

ggplot(data=size_df,aes(sample,read_millions)) +
   geom_bar(stat="identity",fill="indianred",colour="indianred",width=0.7,alpha=0.7)+
   coord_flip()+
   theme_bw()
```

## Cluster samples with hierarchical clustering and display the clustering tree

```{r}
# scale for rows (each gene), to normalize for basal expression level differences
clu_data <- scale(t(filter_counts_df))
# distance matrix (we are clustering samples, not genes)
dd <- dist(clu_data, method = "euclidean") 
hc <- hclust(dd, method="ward.D")
# display dendogram
plot(hc) 

## PCA analysis of samples
data.matrix <- filter_counts_df
data.matrix_t = t(data.matrix)
ordered_matrix_t<- merge(c_anno_df, data.matrix_t, by.y = 'row.names', by.x = "sample")
sorted_ordered_matrix_t <- ordered_matrix_t[order(ordered_matrix_t$condition),]

color <- c(rep("red",50),rep("blue",50))
row.names(sorted_ordered_matrix_t)= sorted_ordered_matrix_t$sample
matrix <- sorted_ordered_matrix_t[1:nrow( sorted_ordered_matrix_t),3:ncol(sorted_ordered_matrix_t)]


data.PC <- prcomp(matrix,scale.=TRUE)
plot(data.PC$x[,1:2],xlim=c(-200,200),ylim=c(-200,200),col=color,pch=19)
text(data.PC$x[,1],data.PC$x[,2]+10,colnames(filter_counts_df),cex=0.7)
```

# create a DGRList object

```{r}
c_anno_df_copy <- c_anno_df
#c_anno_df <- c_anno_df[order(ordered_matrix_t$condition),]
f_matrix <- t(matrix)
edge_c <- DGEList(counts=f_matrix,group=c_anno_df$condition,samples=c_anno_df,genes=filter_anno_df) 
edge_c
```

# normalization with the edgeR package (TMM method)

```{r}
edge_n <- calcNormFactors(edge_c,method="TMM")
edge_n
```

```{r}
cpm_table <- as.data.frame(round(cpm(edge_n),2))
head(cpm_table)
long_cpm_df <- gather(cpm_table, key = "sample", value = "CPM") 

ggplot(data=long_cpm_df,aes(sample,CPM+1)) +
   geom_boxplot(colour="olivedrab",fill="olivedrab",alpha=0.7)+
   theme_bw()+
   scale_y_log10() 
```

```{r}
clu_data <- t(scale(t(cpm_table))) 
dd <- dist(t(clu_data), method = "euclidean") 
hc <- hclust(dd, method="ward.D")
plot(hc) 
```

```{r}
data.matrix <- cpm_table
color <- c(rep("red",50),rep("blue",50))
data.PC <- prcomp(t(data.matrix),scale.=TRUE)
plot(data.PC$x[,1:2],xlim=c(-200,200),ylim=c(-200,200),col=color,pch=19)
text(data.PC$x[,1],data.PC$x[,2]+10,colnames(f_matrix),cex=0.7)
```

```{r}
design <- model.matrix(~0+group, data=edge_n$samples)
colnames(design) <- levels(edge_n$samples$group)
rownames(design) <- edge_n$samples$sample
design
```

```{r}
# calculate dispersion and fit with edgeR (necessary for differential expression analysis)
edge_d <- estimateDisp(edge_n,design)
edge_f <- glmQLFit(edge_d,design) 
```

```{r}
contro <- makeContrasts("case-control", levels=design) 
```

```{r}
# fit the model with generalized linear models
edge_t <- glmQLFTest(edge_f,contrast=contro)
DEGs <- as.data.frame(topTags(edge_t,n=20,p.value = 0.01,sort.by = "logFC"))
DEGs <- as.data.frame(topTags(edge_t,n=20000))
DEGs$class <- "="
DEGs$class[which(DEGs$logCPM>1&DEGs$logFC>1.5&DEGs$FDR<0.01)] = "+"
DEGs$class[which(DEGs$logCPM>1&DEGs$logFC<(-1.5)&DEGs$FDR<0.01)] = "-"
DEGs <- DEGs[order(DEGs$logFC,decreasing = T),]

head(DEGs)
table(DEGs$class)
```

```{r}
## Display an MA plot
input_df <- DEGs
xlabel <- "log2 avg CPM (A)"
ylabel <- "log2 FC case vs control (M)"
par(fig=c(0,1,0,1), mar=c(4,4,1,2), mgp=c(2, 0.75, 0))	
plot(input_df$logCPM, input_df$logFC, xlab=xlabel, ylab=ylabel, 
     col=ifelse(input_df$class=="=","grey70","olivedrab4"), pch=20, frame.plot=TRUE, cex=0.8, main="MA plot")
abline(h=0,lty=2,col="grey20")
```

```{r}
## Display a Volcano plot of the results:
input_df <- DEGs
xlabel <- "log2 FC case vs control"
ylabel <- "-log10 p-value"
par(fig=c(0,1,0,1), mar=c(4,4,1,2), mgp=c(2, 0.75, 0))	
plot(input_df$logFC, -log(input_df$PValue,base=10),xlab=xlabel, ylab=ylabel, 
     col=ifelse(input_df$class=="=","grey70","olivedrab4"), pch=20, frame.plot=TRUE, cex=0.8, main="Volcano plot")
abline(v=0,lty=2,col="grey20")

```

## Heatmap with DEG genes
```{r}
data.matrix <- cpm_table
data.matrix_t = t(data.matrix)

ordered_matrix_t<- merge(c_anno_df, data.matrix_t, by.y = 'row.names', by.x = "sample")
sorted_ordered_matrix_t <- ordered_matrix_t[order(ordered_matrix_t$condition),]
row.names(sorted_ordered_matrix_t)= sorted_ordered_matrix_t$sample
matrix <- sorted_ordered_matrix_t[1:nrow( sorted_ordered_matrix_t),3:ncol(sorted_ordered_matrix_t)]
t_cpm_table <- t(matrix)



cols <- c(rep("chartreuse4",50),rep("burlywood3",50)) 
pal <- c("blue","white","red") 
pal <- colorRampPalette(pal)(100)
sub_matrix <- as.matrix(t_cpm_table[which(rownames(t_cpm_table)%in%DEGs$ensembl_gene_id[which(DEGs$class!="=")]),])
heatmap(sub_matrix, ColSideColors = cols,cexCol = 0.5,margins = c(4,4),col=pal,cexRow = 0.2)
```

```{r}
### Use biomaRt to map Gene symbols, Entrez IDs and Ensembl gen IDs
convert <- getBM(attributes=c("ensembl_gene_id","entrezgene_id","external_gene_name"),
                 filters=c("ensembl_gene_id"), 
                 values=DEGs$ensembl_gene_id,
                 mart = ensembl)
DEGs <- merge(DEGs,convert,by.x="ensembl_gene_id",by.y="ensembl_gene_id")
DEGs <- DEGs[1:nrow(DEGs),1:(ncol(DEGs)-1)]
colnames(DEGs)[2] <- "external_gene_name"
DEGs <- DEGs[which(!is.na(DEGs$entrezgene_id)),] #17741 vs 17355
DEGs <- DEGs[-which(duplicated(DEGs$entrezgene_id)),] # 17275
```

```{r}
## Create a list of upregulated genes
upDEGs <- DEGs %>% filter(class == "+")
```

```{r}
## Perform Gene Ontology enrichment analysis (Biological Process)
ego_BP <- enrichGO(gene = upDEGs$external_gene_name,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "BP",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```

```{r}
## Visualize the top 10 enriched terms with a barplot 
barplot(ego_BP, showCategory = 10)
dotplot(ego_BP, showCategory=10)
heatplot(ego_BP,showCategory =10)
```